### PR TITLE
endpoint: Don't hold the endpoint lock while building

### PR DESCRIPTION
**Pull request by tgraf**

### CODE REVIEW

File: pkg/endpoint/policy.go [hunk 1]
```diff
@@ -309,16 +309,28 @@ func (e *Endpoint) regeneratePolicy(owner Owner) (bool, error) {
...
+	defer e.BuildMutex.Unlock()
+
+	// If endpoint was marked as disconnected then
+	// it won't be regenerated
+	if e.IsDisconnecting() {
+		log.Debugf("Endpoint %d disconnected, skipping build", e.ID)
+		return fmt.Errorf("endpoint disconnected, skipping build")
+	}
+
+	log.Debugf("Regenerating endpoint %d...", e.ID)
```
Comment (ianvernon) [inline pkg/endpoint/policy.go]: Can we have logs with the endpoint ID use log.WIthFields? E.g.:
```
log.WithFields( log.Fields{
                                logfields.EndpointID: ep.StringID(), ...
```

File: pkg/endpoint/policy.go [hunk 2]
```diff
@@ -309,16 +309,28 @@ func (e *Endpoint) regeneratePolicy(owner Owner) (bool, error) {
...
 
 // Called with e.Mutex locked
 func (e *Endpoint) regenerate(owner Owner) error {
+	e.BuildMutex.Lock()
+	defer e.BuildMutex.Unlock()
+
+	// If endpoint was marked as disconnected then
+	// it won't be regenerated
+	if e.IsDisconnecting() {
+		log.Debugf("Endpoint %d disconnected, skipping build", e.ID)
```
Comment (ianvernon) [inline pkg/endpoint/policy.go]: Can we have logs with the endpoint ID use log.WIthFields? E.g.:
```
log.WithFields( log.Fields{
                                logfields.EndpointID: ep.StringID(), ...
```

File: pkg/endpoint/policy.go [hunk 3]
```diff
@@ -386,37 +402,24 @@ func (e *Endpoint) Regenerate(owner Owner) <-chan bool {
...
-			if e.State != StateDisconnected {
-				log.Debugf("Regenerating... [%d]", eID)
-				err = e.regenerate(owner)
-				e.State = StateReady
-			} else {
-				log.Debugf("Endpoint disconnected %d", e.ID)
-				err = fmt.Errorf("Endpoint disconnected")
-			}
-			e.Mutex.Unlock()
+			log.Debugf("Dequeued endpoint %d from build queue", e.ID)
```
Comment (ianvernon) [inline pkg/endpoint/policy.go]: Ditto.

File: pkg/endpoint/policy.go (hunk 2 above)
Comment (tgraf) [reply pkg/endpoint/policy.go]: Good point, I did this before the other PR which added these fields. Fixed.

File: pkg/endpoint/policy.go (hunk 1 above)
Comment (tgraf) [reply pkg/endpoint/policy.go]: Fixed

File: pkg/endpoint/policy.go (hunk 3 above)
Comment (tgraf) [reply pkg/endpoint/policy.go]: Fixed

**Review by ianvernon — APPROVED** [review]
Looks great! The less we block and hold locks, the better. Have you noticed significant performance increases as a result of this change?

**Comment by ianvernon** [comment]
Also, have you ran this with the new deadlock detection? Want to make sure we are covered there.

**Pull request merged**